Database/MySQL/Simple/Arity.hs

```diff
@@ -0,0 +1,49 @@
+{-| Generic calculation of the "arity" of data-types.
+
+This code was originally written by Li-yao Xia. See
+<https://stackoverflow.com/a/56351505/1021134>.
+-}
+{-# LANGUAGE
+    AllowAmbiguousTypes
+  , ScopedTypeVariables
+  , TypeFamilies
+  , UndecidableInstances
+  , FlexibleContexts
+  , DataKinds
+  , TypeOperators
+#-}
+{-# OPTIONS_GHC -Wall -Werror #-}
+
+module Database.MySQL.Simple.Arity
+  ( Arity
+  , arity
+  , KnownNat
+  ) where
+
+import Data.Kind (Type)
+import Data.Proxy (Proxy(Proxy))
+import Numeric.Natural (Natural)
+import GHC.Generics (M1, U1, K1, (:*:), Generic, Rep)
+import GHC.TypeNats (KnownNat, Nat, type (+), natVal)
+
+type family Arity (f :: Type -> Type) :: Nat
+type instance Arity (M1 _ _ f) = Arity f
+type instance Arity (f :*: g) = Arity f + Arity g
+type instance Arity U1 = 0
+type instance Arity (K1 i a) = 1
```

Comment:
Please see my comment on [Xia Li-yao's gist](https://gist.github.com/Lysxia/10f1da354f051b2d2eb24f6aace1bf9c#gistcomment-2929864) for a way to improve the error messages if someone tries to calculate the arity of a type that doesn't have exactly one constructor.

Reply:
That's cool.

Reply:
I added this